### PR TITLE
Automated migrations workflow(optionally SSH, Database URL or API call)

### PR DESCRIPTION
**Pull request by EugeneLightsOn**
- Github workflow to automate alembic migrations.

**AI Description**

This PR introduces automated database migrations using GitHub Actions. It sets up the necessary scripts and configuration files to enable migrations through different methods, such as SSH access to a remote server, connection strings, and Cohere Toolkit's API migrations endpoint.

- Adds a new section in the `.env-template` file to include the `MIGRATE_TOKEN` used for authenticating requests to the migrate endpoint.
- Creates two new shell scripts, `run_docker_migration.sh` and `run_src_migration.sh`, in the `.github/scripts` directory to handle migrations using Docker and source code, respectively.
- Introduces a new GitHub Actions workflow file, `run_alembic_migrations.yml`, which defines the steps for running Alembic migrations using SSH, the migration API, or the database URL.
- Updates the `README.md` file to include a link to a new documentation file, `github_migrations_action.md`, which provides a guide on setting up GitHub Actions for automated database migrations.
- Modifies the `auth.py` file to import the necessary modules for token-based authentication and adds a `verify_migrate_token` function to validate the migration token.
- Updates the `main.py` file to include the `verify_migrate_token` function as a dependency for the `/migrate` endpoint, ensuring that the migration token is verified before applying migrations.

### CODE REVIEW

File: .github/workflows/run_alembic_migrations.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
+name: Automated Migrations
+
```
Comment (tianjing-li) [inline .github/workflows/run_alembic_migrations.yml:2]: Add a comment here mention the 3 types of migrations provided:

- API
- DB
- ssh

File: .github/workflows/run_alembic_migrations.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        SSH_PRIVATE_KEY: ${{ secrets.SSH_PRIVATE_KEY }}
+        SSH_HOST: ${{secrets.SSH_HOST}}
+        SSH_PORT: ${{secrets.SSH_PORT}}
+        SSH_USER_NAME: ${{secrets.SSH_USER_NAME}}
+        DATABASE_URL: ${{secrets.MIGRATION_DATABASE_URL}}
+        MIGRATION_API_ENDPOINT: ${{secrets.MIGRATION_API_ENDPOINT}}
+        MIGRATION_API_TOKEN: ${{secrets.MIGRATION_API_TOKEN}}
+        # Check if the migration API params are present we will use the migration API to run migrations.
+        USE_MIGRATION_API: ${{ (secrets.MIGRATION_API_ENDPOINT != '') && (secrets.MIGRATION_API_TOKEN != '') }}
+        # Check if the database URL is present we will use the database URL to run migrations.
```
Comment (tianjing-li) [inline .github/workflows/run_alembic_migrations.yml:20]: `If database URL is present, run migrations on DB connection`

File: .github/workflows/run_alembic_migrations.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        # Please note that the database connection should be open to the IP from where the migrations are being run.
+        # If you are using a private network, you can set up SSH tunnel to run migrations.
+        USE_DATABASE_URL: ${{ secrets.MIGRATION_DATABASE_URL != '' }}
+        # Check if the SSH params or all params are present we will use SSH to run migrations.
```
Comment (tianjing-li) [inline .github/workflows/run_alembic_migrations.yml:24]: `If SSH params are present, use SSH to run migrations.`

File: .github/workflows/run_alembic_migrations.yml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    environment: production
+    env:
+        SSH_PRIVATE_KEY: ${{ secrets.SSH_PRIVATE_KEY }}
+        SSH_HOST: ${{secrets.SSH_HOST}}
+        SSH_PORT: ${{secrets.SSH_PORT}}
+        SSH_USER_NAME: ${{secrets.SSH_USER_NAME}}
+        DATABASE_URL: ${{secrets.MIGRATION_DATABASE_URL}}
+        MIGRATION_API_ENDPOINT: ${{secrets.MIGRATION_API_ENDPOINT}}
+        MIGRATION_API_TOKEN: ${{secrets.MIGRATION_API_TOKEN}}
+        # Check if the migration API params are present we will use the migration API to run migrations.
```
[the remark this answers is not in the document]
Comment (EugeneLightsOn) [reply .github/workflows/run_alembic_migrations.yml]: Thanks

File: .github/workflows/run_alembic_migrations.yml (hunk 2 above)
Comment (EugeneLightsOn) [reply .github/workflows/run_alembic_migrations.yml]: Done

**Review by tianjing-li — APPROVED** [review]
Approved

**Pull request merged**